### PR TITLE
Modularity matrix

### PR DESCRIPTION
**Pull request by jg-you**
This pull request implements the modularity matrix (MEJ Newman, 2006, PNAS) with associated test suites and spectrum operations.
I did not implement any weighted/multigraph/directed methods since there is no consensus on the definition, to the best of my knowledge.

This is my first contribution to NetworkX, comments are more than welcome.

### CODE REVIEW

**Comment by jg-you** [comment]
Had to use `from __future__ import division`. 
This pull request now works both with `python2.7` and `python3.4`.

File: networkx/linalg/modularitymatrix.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+       If nodelist is None, then the ordering is produced by G.nodes().
+
+    Returns
+    -------
+    B : Numpy matrix
+      The modularity matrix of G.
+
+    Notes
+    -----
+    For MultiGraph/MultiDiGraph, the edges weights are summed.
```
Comment (chebee7i) [inline networkx/linalg/modularitymatrix.py]: Given the `@not_implemented_for` decorators, what does this note mean. Similarly with `modularity_spectrum`?

Comment (jg-you) [inline networkx/linalg/modularitymatrix.py]: Oops, it slipped through (I used `linealg/laplacianmatrix.py` as a base template)
My apologies.

**Comment by Midnighter** [comment]
I'm sorry if you asked for this functionality before, I must have missed it. Both @bjedwards and myself have implemented the modularity maximization (also for the directed case) before. You can find his [GSoC project](https://bitbucket.org/bedwards/networkx-community) and [my own code](https://github.com/Midnighter/Everyday-Utilities/blob/master/meb/utils/network/algorithms.py).

Both are dated and not fully integrated so it's great that you're issuing this pull request. I don't have the time right now to compare the code but maybe it'll be helpful. Cheers.

**Comment by jg-you** [comment]
Thanks for the answer @Midnighter.
I never asked for that feature before, but merely implemented it straight away.
Your code does implement this matrix (line 154 of [https://github.com/Midnighter/Everyday-Utilities/blob/master/meb/utils/network/algorithms.py](your own source code)), although it is strictly in the context of community detection.
[GSoc](https://bitbucket.org/bedwards/networkx-community/src/5f88c6ae0db47fd3a5b4a52988093b448f083ca8/networkx/algorithms/community/spectrum.py?at=default#) does not implement the modularity matrix spectrum, and the modularity matrix strictly appears in [the context of modularity optimization (line 70)](https://bitbucket.org/bedwards/networkx-community/src/5f88c6ae0db47fd3a5b4a52988093b448f083ca8/networkx/algorithms/community/partition.py?at=default#).

The idea behind this pull request is to implement a method that fetches the modularity matrix itself, independent of any optimization methods. 
Much like the Laplacian, this matrix does not necessarily requires community information to be of any use.
It has seen been the subject of a renewed interest, due to papers that focus on matrix spectrum calculations (e.g. http://arxiv.org/abs/1306.2507 or http://arxiv.org/abs/1208.1275).

That being said, I'm not aware of your current development targets / philosophy.
I think that the code would fit the present model where it is, but I might be wrong.

**Comment by chebee7i** [comment]
I like this. Code is clean and it's standalone. Let's merge now, and you can submit another pull request with additional features. The directed extensions would be nice.

**Pull request merged**